src/gui/static/src/app/services/exchange.service.ts

```diff
@@ -0,0 +1,71 @@
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpHeaders } from '@angular/common/http';
+import { Observable } from 'rxjs/Observable';
+import { ExchangeOrder, TradingPair } from '../app.datatypes';
+import { environment } from '../../environments/environment';
+
+@Injectable()
+export class ExchangeService {
+  private readonly API_ENDPOINT = 'https://swaplab.cc/api/v3';
+  private readonly API_KEY = 'w4bxe2tbf9beb72r';
+
+  private lastOrder: ExchangeOrder;
+
+  constructor(
+    private http: HttpClient,
+  ) { }
+
+  tradingPairs(): Observable<TradingPair[]> {
+    return this.post('trading_pairs').map(data => data.result);
+  }
+
+  exchange(pair: string, fromAmount: number, toAddress: string): Observable<ExchangeOrder> {
+    return this.post('orders', { pair, fromAmount, toAddress }).map(data => data.result);
+  }
+
+  status(id: string): Observable<ExchangeOrder> {
+    let isDone = false, shouldContinue = true;
+
+    return Observable
+      .timer(0, 30 * 1000)
+      .flatMap(() => this.post('orders/status', { id }))
+      .map(data => data.result)
+      .takeWhile((order: ExchangeOrder) => {
```

Comment:
If this `takeWhile` is placed right below `timer` and you put the `if (['complete', 'error'].indexOf(…` part inside `map`, the `shouldContinue` variable is no longer necessary, the code would be simpler and it would avoid making an extra connection to the Server after the next timer execution.